### PR TITLE
[WIP] add TPU metrics reduction.

### LINKED ISSUE #965: Metrics reduction on distributed TPU setting
## 🚀 Feature

Ignite will support distributed training on TPU (e.g. #960). Currently, metric's computation is impacted in the same way as for DDP on GPUs.

Idea is to improve metric's computation and reduce internal values as it is done for DDP:
https://github.com/pytorch/ignite/blob/0b363974e93e2f79e468cf4a7851a149b96028a7/ignite/metrics/metric.py#L92

To check if we are running in distributed TPU, we can opt to
```
# global definition
try:
     import torch_xla.core.xla_model as xm
     on_xla_device = True
except ImportError:
     on_xla_device = False
```
and if we need to reduce:
```
xm.xrt_world_size() > 1
```

This issues depends on #963

### PR DESCRIPTION
**Pull request by erip**
Fixes #965 

Description: Adds TPU metrics all_reduce

~TODO: add tests~

Check list:
* [x] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/metric.py [hunk 1]
```diff
@@ -91,14 +98,14 @@ def compute(self) -> Any:
...
-
+    def _do_reduction(
+        self,
+        tensor: Union[torch.Tensor, numbers.Number],
+        reduction: Callable[[Union[torch.Tensor, numbers.Number]], None],
+    ) -> Union[torch.Tensor, numbers.Number]:
         tensor_to_number = False
         if isinstance(tensor, numbers.Number):
-            tensor = torch.tensor(tensor, device=self._device)
+            tensor = torch.tensor(tensor, device=self._device, dtype=torch.float)
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:108]: Passing to `dtype=torch.float` is only necessary for TPU as they do not support `torch.long` and probably others.
Maybe we can wait until XLA fix it and remove casting

File: ignite/metrics/metric.py [hunk 2]
```diff
@@ -108,14 +115,37 @@ def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union
...
+        dist.all_reduce(tensor)
+
+    def _tpu_sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union[torch.Tensor, numbers.Number]:
+        return self._do_reduction(tensor, self._tpu_reduce)
+
+    def _gpu_sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union[torch.Tensor, numbers.Number]:
+        return self._do_reduction(tensor, self._gpu_reduce)
+
+    def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union[torch.Tensor, numbers.Number]:
+        if on_xla_device:
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:140]: IMO, we need to test something equivalent to `dist.is_initialized()` which can be probably done by `xm.xrt_world_size() > 1`. This I checked on Colab.

File: ignite/metrics/metric.py (hunk 1 above)
Comment (erip) [reply ignite/metrics/metric.py]: In the interim I've added an optional dtype arg to the `_do_reduction` method which is only populated in the TPU case.

**Comment by erip** [comment]
Should we wait for #992 to merge this? Since nothing here is user-facing we don't have to worry about breaking this, but it's probably best to hold off. WDYT @vfdev-5?

**Comment by vfdev-5** [comment]
@erip we can refactor a bit current tests to accept `world_size` and `rank` as args and draft some TPU tests. In a PR for #992 we can rewrite the injection with new API.

File: tests/ignite/metrics/test_metric.py [hunk 3]
```diff
@@ -602,37 +605,46 @@ def update(self, output):
     m.compute()
 
 
+@pytest.mark.tpu
+@pytest.mark.skipif(not on_xla_device or xm.xrt_world_size() < 1, reason="Skip if no TPU")
+def test_distrib_gpu():
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_metric.py:611]:
```suggestion
def test_distrib_tpu():
```

Comment (erip) [reply tests/ignite/metrics/test_metric.py:611]: ugh, oops.

File: ignite/metrics/metric.py [hunk 4]
```diff
@@ -108,14 +118,37 @@ def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union
...
-        dist.all_reduce(tensor)
+        reduction(tensor)
 
         if tensor_to_number:
             return tensor.item()
         return tensor
 
+    @staticmethod
+    def _tpu_reduce(tensor: Union[torch.Tensor, numbers.Number]) -> None:
+        xm.all_reduce("sum", [tensor,], torch.float)
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:129]: I'm not sure about `torch.float` here. The API is a bit different
http://pytorch.org/xla/release/1.5/index.html#torch_xla.core.xla_model.all_reduce

Comment (erip) [reply ignite/metrics/metric.py:129]: oops, this was a mistake when cleaning up the abstraction. that's supposed to happen in the number/tensor conversion.

**Comment by erip** [comment]
It looks like the remaining tests are hanging (probably from normal flakiness). Does GitHub actions provide you with the option to rerun individual suites manually, @vfdev-5?

**Comment by vfdev-5** [comment]
@erip I can only "cancel" and "re-run all jobs". It should be distributed tests...

File: ignite/metrics/metric.py [hunk 5]
```diff
@@ -108,14 +118,37 @@ def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union
...
+    def _gpu_sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union[torch.Tensor, numbers.Number]:
+        return self._do_reduction(tensor, self._gpu_reduce)
+
+    def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union[torch.Tensor, numbers.Number]:
+        if not (dist.is_available() and dist.is_initialized()) and not xrt_world_size > 0:
+            # Nothing to reduce
+            return tensor
+
+        if on_xla_device:
+            return self._tpu_sync_all_reduce(tensor)
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:148]: Can we directly call `self._do_reduction(tensor, self._gpu_reduce)` intead ?
The purpose of `_gpu_sync_all_reduce` and `_tpu_sync_all_reduce` is very limited for me...

**Comment by vfdev-5** [comment]
@erip can add more tests with real metrics: Accuracy, Recall, Precision etc ?

**Comment by erip** [comment]
Absolutely.

**Comment by erip** [comment]
The distributed test fixtures are going to require some clever thought. [`xmp.spawn`](https://pytorch.org/xla/release/1.5/index.html#torch_xla.distributed.xla_multiprocessing.spawn) requires a function that is passed to all workers. In the case of our specs, that would be the actual test to be run by each worker and then joined. I'm trying to figure out how to defer that info until the tests run since it won't be available in the `conftest.py` where fixtures are set up.

**Comment by vfdev-5** [comment]
@erip yes, let's execute tpu tests with a single worker which does `xmp.spawn` etc. In fixtures we can provide an executing method that requires test as input function.

File: tests/ignite/metrics/test_precision.py [hunk 6]
```diff
@@ -730,7 +729,7 @@ def _test(average, n_epochs):
         n_classes = 7
 
         offset = n_iters * s
-        y_true = torch.randint(0, n_classes, size=(offset * dist.get_world_size(),)).to(device)
+        y_true = torch.randint(0, n_classes, size=(offset * world_size)).to(device)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_precision.py:732]:
```suggestion
        y_true = torch.randint(0, n_classes, size=(offset * world_size, )).to(device)
```

**Comment by erip** [comment]
> @erip yes, let's execute tpu tests with a single worker which does `xmp.spawn` etc. In fixtures we can provide an executing method that requires test as input function.

So your thought is something like...

```python
@pytest.mark.tpu
@pytest.mark.distributed
@pytest.mark.skipif(..., reason="Skip if no TPU")
def test_distrib_gpu(local_rank, distributed_context_single_node_xrt_executor):
    device = xm.get_device()
    args = (device, xm.get_ordinal(), xm.get_world_size())
    distributed_context_single_node_xrt_executor(_test_distrib_itegration_multiclass, args)
    distributed_context_single_node_xrt_executor(_test_distrib_itegration_multilabel, args)
```

or did I misunderstand?

**Comment by vfdev-5** [comment]
> > @erip yes, let's execute tpu tests with a single worker which does `xmp.spawn` etc. In fixtures we can provide an executing method that requires test as input function.
> 
> So your thought is something like...
> 
> ```python
> @pytest.mark.tpu
> @pytest.mark.distributed
> @pytest.mark.skipif(..., reason="Skip if no TPU")
> def test_distrib_tpu(local_rank, distributed_context_single_node_xrt_executor):
>     args = (xm.get_device(), xm.get_ordinal(), xm.get_world_size())
>     distributed_context_single_node_xrt_executor(_test_distrib_itegration_multiclass, args)
>     distributed_context_single_node_xrt_executor(_test_distrib_itegration_multilabel, args)
> ```
> 
> or did I misunderstand?

Yes, something like that :)

File: ignite/metrics/metric.py [hunk 7]
```diff
@@ -108,14 +118,31 @@ def _sync_all_reduce(self, tensor: Union[torch.Tensor, numbers.Number]) -> Union
...
             return tensor.item()
         return tensor
 
+    @staticmethod
+    def _tpu_reduce(tensor: Union[torch.Tensor, numbers.Number]) -> None:
+        xm.all_reduce("sum", [tensor,])
+
+    @staticmethod
+    def _gpu_reduce(tensor: Union[torch.Tensor, numbers.Number]) -> None:
+        dist.barrier()
```
Comment (sdesrozis) [inline ignite/metrics/metric.py:133]: I think this barrier is useless. The following all_reduce() is a sync collective call.

Comment (erip) [reply ignite/metrics/metric.py:133]: Since it's there currently, I'd be a bit wary of changing the behavior in this PR. Since a lot of things are changing, I'd hate to have less information about potential breakages.

Comment (sdesrozis) [reply ignite/metrics/metric.py:133]: I agree!! 😊

**Comment by erip** [comment]
This is going to be blocked until https://github.com/pytorch/xla/pull/2023 is merged, I think.

**Comment by vfdev-5** [comment]
On nightly the problem can be found here : https://pipelines.actions.githubusercontent.com/GjXvbhyRrTHRIrk9p03R8DrZ91xWSv51MZra799xsxhclrqZfD/_apis/pipelines/1/runs/934/signedlogcontent/4?urlExpires=2020-05-04T21%3A17%3A13.3601076Z&urlSigningMethod=HMACV1&urlSignature=PSs%2Bazok0xClsvtr8MlEGCgxKYN7mJQjgyXt60%2FLKOE%3D

A warning is not raised for multilabel case without average...

**Comment by erip** [comment]
Building the nightly xla image now.

**Comment by vfdev-5** [comment]
@erip the problem more precisely is here : https://github.com/pytorch/ignite/blob/ecf10c61f95baa3ae8421ddde187423ed9c251fc/ignite/metrics/precision.py#L22

**Comment by erip** [comment]
Note for myself: need more tests. I believe many of them won't run on CI, but we can test on colab or kaggle.

File: ignite/metrics/metric.py [hunk 8]
```diff
@@ -11,6 +11,15 @@
 
 from ignite.engine import Events, Engine
 
+try:
+    import torch_xla.core.xla_model as xm
+
+    on_xla_device = True
+    xrt_world_size = xm.xrt_world_size()
```
Comment (vfdev-5) [inline ignite/metrics/metric.py:18]: By calling `xm.xrt_world_size()` outside `xm.spawn` we will always have 1.

**Comment by vfdev-5** [comment]
@erip I changed the base branch from master to idist. As now in idist we have some common dist handling we can adapt metrics reduction tests to use idist instead of torch native dist

**Comment by erip** [comment]
That's a good thought. Thanks - will try to roll in some of the new APIs

**Comment by vfdev-5** [comment]
@erip I'll try to do a merge with idist branch

**Comment by vfdev-5** [comment]
Closed in favor of #1042
